compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUDistribute.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMGPU/KernelConfig.h"
+#include "iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Codegen/Transforms/Transforms.h"
+#include "mlir/Conversion/FuncToLLVM/ConvertFuncToLLVM.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/Transforms/Passes.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Support/MathExtras.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/SideEffectUtils.h"
+
+#define DEBUG_TYPE "iree-llvmgpu-distribute"
+
+namespace mlir {
+namespace iree_compiler {
+
+/// Clean up barriers if we have no shared memory allocations we expect to not
+/// need any barriers and remove them.
+static void cleanUpBarriers(func::FuncOp funcOp) {
+  bool hasAlloc = false;
+  SmallVector<Operation*> barriers;
+  funcOp.walk([&](Operation* op) {
+    if (isa<memref::AllocOp>(op))
```

Comment:
something with side effects rather than hardcode the op ?

Reply:
I don't think it is possible at this point since there will be ops with allocation side effects (the hal ops). This is there to avoid perf regressions but this is only correct based on assumption in the rest of iree flow